docker/utils/utils.py

```diff
@@ -128,7 +128,13 @@ def exclude_paths(root, patterns, dockerfile=None):
     paths = get_paths(root, exclude_patterns, include_patterns,
                       has_exceptions=len(exceptions) > 0)
 
-    return set(paths)
+    return set(paths).union(
+        # If the Dockerfile is in a subdirectory that is excluded, get_paths
+        # will not descend into it and the file will be skipped. This ensures
+        # it doesn't happen.
+        set([dockerfile])
+        if os.path.exists(os.path.join(root, dockerfile)) else {}
```

Comment:
If utils.tar() is called on a folder that doesn't contain a Dockerfile, it would break without this change. It doesn't affect docker-py directly I think (we have some tar unit tests that break), but it's part of our public API and some people seem to use it: https://github.com/docker/docker-py/pull/865